File: reference/python/ssklib/gltf.py
"""Write glTF 2.0 output (.glb and .gltf)."""

import json
import os
import struct

import numpy as np

# public API

def write_glb(vertices: np.ndarray, faces: np.ndarray, path: str):

    verts, norms, idx = _unindex(
        np.asarray(vertices, dtype=np.float32),
        np.asarray(faces, dtype=np.uint32))

    buf = _buffer(verts, norms, idx)
    g = _gltf_json(verts, norms, idx, len(buf))

    jb = json.dumps(g, separators=(',', ':')).encode('utf-8')
    jb += b' ' * ((-len(jb)) % 4)

    total = 12 + 8 + len(jb) + 8 + len(buf)
    with open(path, 'wb') as f:
        f.write(struct.pack('<III', 0x46546C67, 2, total))   # header
        f.write(struct.pack('<II', len(jb), 0x4E4F534A))     # JSON chunk
        f.write(jb)
        f.write(struct.pack('<II', len(buf), 0x004E4942))    # BIN chunk
        f.write(buf)


def write_gltf(vertices: np.ndarray, faces: np.ndarray, path: str):

    verts, norms, idx = _unindex(
        np.asarray(vertices, dtype=np.float32),
        np.asarray(faces, dtype=np.uint32))

    buf = _buffer(verts, norms, idx)

    bin_name = os.path.splitext(os.path.basename(path))[0] + '.bin'
    bin_path = os.path.join(os.path.dirname(path) or '.', bin_name)
    g = _gltf_json(verts, norms, idx, len(buf), bin_uri=bin_name)

    with open(bin_path, 'wb') as f:
        f.write(buf)
    with open(path, 'w', encoding='utf-8') as f:
        json.dump(g, f, indent=2)

# internals

def _unindex(verts, faces):

    v0, v1, v2 = verts[faces[:, 0]], verts[faces[:, 1]], verts[faces[:, 2]]
    n = np.cross(v1 - v0, v2 - v0).astype(np.float32)
    lengths = np.linalg.norm(n, axis=1, keepdims=True)
    n /= np.where(lengths < 1e-10, 1.0, lengths)

    nf = len(faces)
    ov = np.empty((nf * 3, 3), dtype=np.float32)
    on = np.empty((nf * 3, 3), dtype=np.float32)
    ov[0::3], ov[1::3], ov[2::3] = v0, v1, v2
    on[0::3] = on[1::3] = on[2::3] = n
    of = np.arange(nf * 3, dtype=np.uint32).reshape(-1, 3)
    return ov, on, of


def _buffer(verts, norms, faces):
    b = verts.tobytes() + norms.tobytes() + faces.tobytes()
    b += b'\x00' * ((-len(b)) % 4)
    return b


def _gltf_json(verts, norms, faces, buf_len, bin_uri=None):
    nv, nt = len(verts), len(faces)
    pl, nl, il = nv * 12, nv * 12, nt * 12

    buf_entry = {'byteLength': buf_len}
    if bin_uri is not None:
        buf_entry['uri'] = bin_uri

    return {
        'asset': {'version': '2.0', 'generator': 'ssk'},
        'scene': 0,
        'scenes': [{'nodes': [0]}],
        'nodes': [{'mesh': 0}],
        'meshes': [{'primitives': [{
            'attributes': {'POSITION': 0, 'NORMAL': 1},
            'indices': 2, 'mode': 4,
        }]}],
        'accessors': [
            {'bufferView': 0, 'byteOffset': 0, 'componentType': 5126,
             'count': nv, 'type': 'VEC3',
             'min': verts.min(axis=0).tolist(),
             'max': verts.max(axis=0).tolist()},
            {'bufferView': 1, 'byteOffset': 0, 'componentType': 5126,
             'count': nv, 'type': 'VEC3'},
            {'bufferView': 2, 'byteOffset': 0, 'componentType': 5125,
             'count': nt * 3, 'type': 'SCALAR'},
        ],
        'bufferViews': [
            {'buffer': 0, 'byteOffset': 0,      'byteLength': pl, 'target': 34962},
            {'buffer': 0, 'byteOffset': pl,      'byteLength': nl, 'target': 34962},
            {'buffer': 0, 'byteOffset': pl + nl, 'byteLength': il, 'target': 34963},
        ],
        'buffers': [buf_entry],
    }

File: reference/typescript/src/gltf.ts
import type { GltfOutput, MeshData } from './types.js';

const GLB_MAGIC = 0x46546c67;
const JSON_CHUNK = 0x4e4f534a;
const BIN_CHUNK = 0x004e4942;

export function writeGlb(mesh: MeshData): Uint8Array {
  const unindexed = unindex(mesh);
  const bin = buffer(unindexed.vertices, unindexed.normals, unindexed.faces);
  const jsonBytes = paddedJson(gltfJson(unindexed.vertices, unindexed.normals, unindexed.faces, bin.length));

  const total = 12 + 8 + jsonBytes.length + 8 + bin.length;
  const out = new Uint8Array(total);
  const view = new DataView(out.buffer);
  let offset = 0;
  view.setUint32(offset, GLB_MAGIC, true); offset += 4;
  view.setUint32(offset, 2, true); offset += 4;
  view.setUint32(offset, total, true); offset += 4;
  view.setUint32(offset, jsonBytes.length, true); offset += 4;
  view.setUint32(offset, JSON_CHUNK, true); offset += 4;
  out.set(jsonBytes, offset); offset += jsonBytes.length;
  view.setUint32(offset, bin.length, true); offset += 4;
  view.setUint32(offset, BIN_CHUNK, true); offset += 4;
  out.set(bin, offset);
  return out;
}

export function writeGltf(mesh: MeshData, binUri = 'model.bin'): GltfOutput {
  const unindexed = unindex(mesh);
  const bin = buffer(unindexed.vertices, unindexed.normals, unindexed.faces);
  return { json: gltfJson(unindexed.vertices, unindexed.normals, unindexed.faces, bin.length, binUri), bin, binUri };
}

function unindex(mesh: MeshData): MeshData & { normals: number[][] } {
  const vertices: number[][] = [];
  const normals: number[][] = [];
  const faces: number[][] = [];
  mesh.faces.forEach((face, faceIndex) => {
    const v0 = mesh.vertices[face[0]];
    const v1 = mesh.vertices[face[1]];
    const v2 = mesh.vertices[face[2]];
    const n = normal(v0, v1, v2);
    const base = faceIndex * 3;
    vertices.push(v0, v1, v2);
    normals.push(n, n, n);
    faces.push([base, base + 1, base + 2]);
  });
  return { vertices, normals, faces };
}

function normal(v0: number[], v1: number[], v2: number[]): number[] {
  const a = [v1[0] - v0[0], v1[1] - v0[1], v1[2] - v0[2]];
  const b = [v2[0] - v0[0], v2[1] - v0[1], v2[2] - v0[2]];
  const n = [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]];
  const length = Math.hypot(n[0], n[1], n[2]);
  return length < 1e-10 ? [0, 0, 0] : [n[0] / length, n[1] / length, n[2] / length];
}

function buffer(vertices: number[][], normals: number[][], faces: number[][]): Uint8Array {
  const vertexBytes = float32Bytes(vertices.flat());
  const normalBytes = float32Bytes(normals.flat());
  const indexBytes = uint32Bytes(faces.flat());
  const length = align4(vertexBytes.length + normalBytes.length + indexBytes.length);
  const out = new Uint8Array(length);
  let offset = 0;
  out.set(vertexBytes, offset); offset += vertexBytes.length;
  out.set(normalBytes, offset); offset += normalBytes.length;
  out.set(indexBytes, offset);
  return out;
}

function gltfJson(vertices: number[][], normals: number[][], faces: number[][], bufLen: number, binUri?: string): unknown {
  const nv = vertices.length;
  const nt = faces.length;
  const pl = nv * 12;
  const nl = normals.length * 12;
  const il = nt * 12;
  const bufferEntry: Record<string, unknown> = { byteLength: bufLen };
  if (binUri) bufferEntry.uri = binUri;
  return {
    asset: { version: '2.0', generator: 'ssk' },
    scene: 0,
    scenes: [{ nodes: [0] }],
    nodes: [{ mesh: 0 }],
    meshes: [{ primitives: [{ attributes: { POSITION: 0, NORMAL: 1 }, indices: 2, mode: 4 }] }],
    accessors: [
      { bufferView: 0, byteOffset: 0, componentType: 5126, count: nv, type: 'VEC3', min: min(vertices), max: max(vertices) },
      { bufferView: 1, byteOffset: 0, componentType: 5126, count: nv, type: 'VEC3' },
      { bufferView: 2, byteOffset: 0, componentType: 5125, count: nt * 3, type: 'SCALAR' },
    ],
    bufferViews: [
      { buffer: 0, byteOffset: 0, byteLength: pl, target: 34962 },
      { buffer: 0, byteOffset: pl, byteLength: nl, target: 34962 },
      { buffer: 0, byteOffset: pl + nl, byteLength: il, target: 34963 },
    ],
    buffers: [bufferEntry],
  };
}

function float32Bytes(values: number[]): Uint8Array {
  const bytes = new Uint8Array(values.length * 4);
  const view = new DataView(bytes.buffer);
  values.forEach((value, index) => view.setFloat32(index * 4, value, true));
  return bytes;
}

function uint32Bytes(values: number[]): Uint8Array {
  const bytes = new Uint8Array(values.length * 4);
  const view = new DataView(bytes.buffer);
  values.forEach((value, index) => view.setUint32(index * 4, value, true));
  return bytes;
}

function paddedJson(value: unknown): Uint8Array {
  const raw = new TextEncoder().encode(JSON.stringify(value));
  const out = new Uint8Array(align4(raw.length));
  out.set(raw);
  out.fill(0x20, raw.length);
  return out;
}

function align4(value: number): number {
  return value + ((4 - value % 4) % 4);
}

function min(values: number[][]): number[] {
  return [0, 1, 2].map((axis) => Math.min(...values.map((value) => value[axis])));
}

function max(values: number[][]): number[] {
  return [0, 1, 2].map((axis) => Math.max(...values.map((value) => value[axis])));
}
